sequence_processing_pipeline/Pipeline.py

```diff
@@ -12,11 +12,79 @@
 import sample_sheet
 import pandas as pd
 from collections import defaultdict
+from datetime import datetime
+import pytz
+from xml.etree import ElementTree as ET
 
 
 logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
 
 
+class InstrumentUtils():
+    types = {'A': 'NovaSeq 6000', 'D': 'HiSeq 2500', 'FS': 'iSeq',
+             'K': 'HiSeq 4000', 'LH': 'NovaSeq X Plus', 'M': 'MiSeq',
+             'MN': 'MiniSeq',
+             # SN – RapidRun which is HiSeq 2500
+             'SN': 'RapidRun'}
+
+    @staticmethod
+    def get_instrument_type(run_id):
+        tmp = run_id.split('_')
+
+        # remove digits from substring to get instrument code.
+        code = ''.join([i for i in tmp[1] if not i.isdigit()])
+        try:
+            return InstrumentUtils.types[code]
+        except KeyError:
+            raise ValueError(f"Instrument code '{code}' is of unknown type")
+
+    @staticmethod
+    def get_instrument_id(run_id):
+        tmp = run_id.split('_')
+
+        # remove letters from substring to get instrument id.
+        return ''.join([i for i in tmp[1] if i.isdigit()])
+
+    @staticmethod
+    def get_date(run_id):
+        ds = run_id.split('_')[0]
+
+        if len(ds) == 6:
+            year = int(ds[0:2]) + 2000
+            month = int(ds[2:4])
+            date = int(ds[4:6])
+        elif len(ds) == 8:
+            year = int(ds[0:4])
+            month = int(ds[4:6])
+            date = int(ds[6:8])
+        else:
+            raise ValueError("Couldn't parse datestamp: '%s'" % ds)
+
+        if month < 1 or month > 12:
+            raise ValueError("Invalid month: '%s'" % month)
+
+        if date < 1 or date > 31:
+            raise ValueError("Invalid date: '%s'" % date)
+
+        return datetime(year, month,
+                        date, tzinfo=pytz.timezone("America/Los_Angeles"))
+
+    @staticmethod
+    def get_flow_cell_mode(run_directory):
```

Comment:
this is never used, right? Anyway, what's the plan of implementation?

Reply:
Right we're not using it right now and I'm not sure who would use it. However, extracting it was discussed in #106 so felt it was important to include it. Gail has replied that iSeq and MiSeq use a standard flow-cell type and Jeff has included some other possibilities I can look into.

We don't need this at the moment but we do need to identify the instrument type so we can supply the correct resource allocations for a job. Do you all think we should split this into two separate PRs?